### PR TITLE
Add 1LHAASO to gammapy.catalog

### PR DESCRIPTION
**Pull request by QRemy**
Add 1LHAASO catalog class.
This requires to add the catalog FITS file to gammapy-data.

### CODE REVIEW

File: gammapy/catalog/lhaaso.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+        """
+
+        if which == "both":
+            wcda = self.sky_model(which="WCDA")
+            km2a = self.sky_model(which="KM2A")
+            models = [m for m in [wcda, km2a] if m is not None]
+            if len(models) == 2:
+                geom = self._get_components_geom(models)
+                mask = geom.energy_mask(energy_max=25 * u.TeV)
```
Comment (bkhelifi) [inline gammapy/catalog/lhaaso.py:129]: The hardcoded value of energy threshold is applied applied to both sub-arrays. In the paper, it seems that the threshold is 1TeV for WCDA (with Eref=3TeV) and 25TeV for  KM2A (with Eref=50TeV).
In any case, these hardcoded values are not so great and one could imagine to save them into meta data for the FITS file.

File: gammapy/catalog/lhaaso.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        regions = [m.spatial_model.evaluation_region for m in models]
+        geom = RegionGeom.from_regions(
+            regions, binsz_wcs="0.05 deg", axes=[energy_axis]
+        )
+        return geom.to_wcs_geom()
+
+    def sky_model(self, which="both"):
+        """Sky model (`~gammapy.modeling.models.SkyModel`).
+
+        * ``which="both"`` - Use first model or create a composite template if both models are availble
```
Comment (bkhelifi) [inline gammapy/catalog/lhaaso.py:117]:
```suggestion
        * ``which="both"`` - Use first model or create a composite template if both models are available.
```

File: gammapy/catalog/lhaaso.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        Entries not repeated for the second analysis are taken from the first one.
+        """
+
+        lat_0 = self._get("DECJ2000", which)
+        pars = {"lon_0": self._get("RAJ2000", which), "lat_0": lat_0, "frame": "fk5"}
+
+        pos_err = self._get("pos_err", which)
+        errs = {
+            "lat_0": pos_err,
+            "lon_0": pos_err / np.cos(lat_0),
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:88]: in degree, why ?

File: gammapy/catalog/lhaaso.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/catalog/lhaaso.py]: If both are available they are converted to a template that uses WCDA below 25 TeV and KM2A above.
For which source do you see the behaviour you describe ?

File: gammapy/catalog/lhaaso.py [hunk 4]
```diff
@@ -20,27 +22,27 @@ class SourceCatalogObject1LHAASO(SourceCatalogObject):
...
     def _get(self, name, which):
         value, _ = self._parse(name, which)
         return value
 
-    def spectral_model(self, which="all"):
+    def spectral_model(self, which="both"):
         """Spectral model (`~gammapy.modeling.models.PowerLawSpectralModel`).
 
-        * ``which="all"`` - First sky model listed for the source.
+        * ``which="both"`` - Use first model or create a composite template if both models are availble
```
Comment (bkhelifi) [inline gammapy/catalog/lhaaso.py:45]:
```suggestion
        * ``which="both"`` - Use first model or create a composite template if both models are available.
```

File: gammapy/catalog/lhaaso.py [hunk 5]
```diff
@@ -66,10 +68,10 @@ def spectral_model(self, which="all"):
 
         return model
 
-    def spatial_model(self, which="all"):
+    def spatial_model(self, which="both"):
         """Spatial model (`~gammapy.modeling.models.SpatialModel`).
 
-        * ``which="all"`` - First sky model listed for the source.
+        * ``which="both"`` - Use first model or create a composite template if both models are availble
```
Comment (bkhelifi) [inline gammapy/catalog/lhaaso.py:74]:
```suggestion
        * ``which="both"`` - Use first model or create a composite template if both models are available.
```

File: gammapy/catalog/lhaaso.py [hunk 6]
```diff
@@ -98,21 +100,51 @@ def spatial_model(self, which="all"):
...
+        )
+        return geom.to_wcs_geom()
+
+    def sky_model(self, which="both"):
         """Sky model (`~gammapy.modeling.models.SkyModel`).
 
-        * ``which="all"`` - First sky model listed for the source.
-        * ``which="KM2A"`` - Sky model for KM2A analysis only.
-        * ``which="WCDA"`` - Sky model for WCDA analysis only.
+        * ``which="both"`` - Use first model or create a composite template if both models are availble
```
Comment (bkhelifi) [inline gammapy/catalog/lhaaso.py]:
```suggestion
        * ``which="both"`` - Use first model or create a composite template if both models are available.
```

File: gammapy/catalog/lhaaso.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkhelifi) [reply gammapy/catalog/lhaaso.py:88]: When lon/lat are read from _get and _parse, a quantity is returned. So, if they are angles, it works. So no problem. But, in general, should we check the consistency of units?

[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:88]: I will add units checks in the tests

File: gammapy/catalog/lhaaso.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        * ``which="KM2A"`` - Sky model for KM2A analysis if available.
+        * ``which="WCDA"`` - Sky model for WCDA analysis if available.
+
+        """
+
+        if which == "both":
+            wcda = self.sky_model(which="WCDA")
+            km2a = self.sky_model(which="KM2A")
+            models = [m for m in [wcda, km2a] if m is not None]
+            if len(models) == 2:
```
Comment (AtreyeeS) [inline gammapy/catalog/lhaaso.py:130]: there is no overlap in energy between the WCDA and KM2A ?

File: gammapy/catalog/lhaaso.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    def __init__(self, filename="$GAMMAPY_DATA/catalogs/1LHAASO_catalog.fits"):
+        table = Table.read(make_path(filename))
+
+        source_name_key = "Source_Name"
+
+        super().__init__(table=table, source_name_key=source_name_key)
+
+    def to_models(self, **kwargs):
+        """Create Models object from catalogue"""
```
Comment (AtreyeeS) [inline gammapy/catalog/lhaaso.py:184]: Would be good to expose `which` option here

**Review by AtreyeeS — COMMENTED** [review]
Thanks @QRemy ! I propose we merge this now, and add more support by the next release

File: gammapy/catalog/lhaaso.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        value, _ = self._parse(name, which)
+        return value
+
+    def spectral_model(self, which):
+        """Spectral model (`~gammapy.modeling.models.PowerLawSpectralModel`).
+
+        * ``which="KM2A"`` - Sky model for KM2A analysis only.
+        * ``which="WCDA"`` - Sky model for WCDA analysis only.
+
+        If only a limit is given for a parameter it is used as value.
```
Comment (QRemy) [inline gammapy/catalog/lhaaso.py:51]:
```suggestion
```

File: gammapy/catalog/lhaaso.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+        return model
+
+    def spatial_model(self, which):
+        """Spatial model (`~gammapy.modeling.models.SpatialModel`).
+
+        * ``which="KM2A"`` - Sky model for KM2A analysis only.
+        * ``which="WCDA"`` - Sky model for WCDA analysis only.
+
+        If only a limit is given for a parameter it is used as value.
```
Comment (QRemy) [inline gammapy/catalog/lhaaso.py:79]:
```suggestion
```

File: gammapy/catalog/lhaaso.py (hunk 7 above)
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:130]: no

File: gammapy/catalog/lhaaso.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    source_object_class = SourceCatalogObject1LHAASO
+
+    def __init__(self, filename="$GAMMAPY_DATA/catalogs/1LHAASO_catalog.fits"):
+        table = Table.read(make_path(filename))
+
+        source_name_key = "Source_Name"
+
+        super().__init__(table=table, source_name_key=source_name_key)
+
+    def to_models(self, **kwargs):
```
Comment (QRemy) [inline gammapy/catalog/lhaaso.py:183]:
```suggestion
    def to_models(self,  which="both"):
```

File: gammapy/catalog/lhaaso.py (hunk 8 above)
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:182]: done

File: gammapy/catalog/lhaaso.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/catalog/lhaaso.py]: removed

**Review by bkhelifi — DISMISSED** [review]
Thanks again, @QRemy

File: gammapy/catalog/lhaaso.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+        lat_0 = self._get("DECJ2000", which)
+        pars = {"lon_0": self._get("RAJ2000", which), "lat_0": lat_0, "frame": "fk5"}
+
+        pos_err = self._get("pos_err", which)
+        scale_r95 = Gauss2DPDF().containment_radius(0.95)
+
+        errs = {
+            "lat_0": pos_err / scale_r95,
+            "lon_0": pos_err / scale_r95 / np.cos(lat_0),
```
Comment (registerrier) [inline gammapy/catalog/lhaaso.py:89]: why is the division by cos(lat) necessary?

File: gammapy/catalog/lhaaso.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        r39, is_ul = self._parse("r39", which)
+        if not is_ul:
+            pars["sigma"] = r39
+            model = Model.create("GaussianSpatialModel", "spatial", **pars)
+        else:
+            model = Model.create("PointSpatialModel", "spatial", **pars)
```
Comment (registerrier) [inline gammapy/catalog/lhaaso.py:97]: Is it better to use a `PointSpatialModel` or a `Gaussian` one with very small `sigma`?

**Review by registerrier — APPROVED** [review]
Thanks @QRemy . This looks good. I have left a few minor inline comments.questions.

File: gammapy/catalog/lhaaso.py (hunk 12 above)
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:89]: not sure it is the same for 4FGL and 3HWC catalogue.

File: gammapy/catalog/lhaaso.py (hunk 13 above)
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:97]: I would say a `PointSpatialModel ` so we don't have to decide about the sigma to use

**Review by AtreyeeS — DISMISSED** [review]
Thanks!

**Review by bkhelifi — APPROVED** [review]
Thanks @QRemy

File: gammapy/catalog/lhaaso.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        value, _ = self._parse(name, which)
+        return value
+
+    def spectral_model(self, which):
+        """Spectral model (`~gammapy.modeling.models.PowerLawSpectralModel`).
+
+        * ``which="KM2A"`` - Sky model for KM2A analysis only.
+        * ``which="WCDA"`` - Sky model for WCDA analysis only.
+
+        Entries not repeated for the second analysis are taken from the first one.
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/catalog/lhaaso.py:51]: yes i removed this behaviour but not the docstring, modified

Comment (QRemy) [inline gammapy/catalog/lhaaso.py:51]:
```suggestion
        Entries not repeated for the second analysis are taken from the first one.
```

File: gammapy/catalog/lhaaso.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+        return model
+
+    def spatial_model(self, which):
+        """Spatial model (`~gammapy.modeling.models.SpatialModel`).
+
+        * ``which="KM2A"`` - Sky model for KM2A analysis only.
+        * ``which="WCDA"`` - Sky model for WCDA analysis only.
+
+        Entries not repeated for the second analysis are taken from the first one.
```
Comment (QRemy) [inline gammapy/catalog/lhaaso.py:78]:
```suggestion
```

File: gammapy/catalog/lhaaso.py (hunk 12 above)
Comment (registerrier) [reply gammapy/catalog/lhaaso.py:87]: I see that this was introduced after issue #2460 . I suppose this implies that the error in the table is given as a proper angular separation and one wants to correct for that. Does this apply all the time?

**Pull request merged**